pyccel/codegen/printing/fcode.py

```diff
@@ -604,47 +602,128 @@ def _print_PythonSum(self, expr):
+        a_code = self._print(expr.a)
+        b_code = self._print(expr.b)
 
-    def _print_NumpyCross(self, expr):
-        return expr.fprint(self._print)
+        if expr.a.order and expr.b.order:
+            if expr.a.order != expr.b.order:
+                raise NotImplementedError("Mixed order matmul not supported.")
+
+        # Fortran ordering
+        if expr.a.order == 'F':
+            return 'matmul({0},{1})'.format(a_code, b_code)
+
+        # C ordering
+        return 'matmul({1},{0})'.format(a_code, b_code)
+
+    def _print_NumpyEmpty(self, expr):
+        errors.report(FORTRAN_ALLOCATABLE_IN_EXPRESSION, symbol=expr, severity='fatal')
 
     def _print_NumpyNorm(self, expr):
-        return expr.fprint(self._print)
+        """Fortran print."""
+
+        if expr.dim:
+            rhs = 'Norm2({},{})'.format(self._print(expr.arg),self._print(expr.dim))
+        else:
+            rhs = 'Norm2({})'.format(self._print(expr.arg))
+
+        return rhs
 
     def _print_NumpyLinspace(self, expr):
-        return expr.fprint(self._print)
+
+        template = '[({start} + {index}*{step},{index} = {zero},{end})]'
+
+        init_value = template.format(
+            start = self._print(expr.start),
+            step  = self._print(expr.step ),
+            index = self._print(expr.index),
+            zero  = self._print(LiteralInteger(0)),
+            end   = self._print(PyccelMinus(expr.size, LiteralInteger(1))),
+        )
+        code = init_value
+
+        return code
+
+    def _print_NumpyWhere(self, expr):
+
+        ind   = self._print(expr.index)
+        mask  = self._print(expr.mask)
+
+        stmt  = 'pack([({ind},{ind}=0,size({mask})-1)],{mask})'.format(ind=ind,mask=mask)
+
+        return stmt
 
     def _print_NumpyArray(self, expr):
-        return expr.fprint(self._print)
+        """Fortran print."""
+
+        # Always transpose indices because Numpy initial values are given with
+        # row-major ordering, while Fortran initial values are column-major
+        shape = expr.shape[::-1]
+
+        # Construct right-hand-side code
+        if expr.rank > 1:
+            arg = functools.reduce(operator.concat, expr.arg)
+            rhs_code = 'reshape({array}, {shape})'.format(
+                    array=self._print(arg), shape=self._print(Tuple(*shape)))
+        else:
+            rhs_code = self._print(expr.arg)
+
+        # If Numpy array is stored with column-major ordering, transpose values
+        if expr.order == 'F' and expr.rank > 1:
+            rhs_code = 'transpose({})'.format(rhs_code)
+
+        return rhs_code
 
     def _print_NumpyFloor(self, expr):
         result_code = self._print_MathFloor(expr)
         return 'real({}, {})'.format(result_code, iso_c_binding["real"][8])
 
     # ======================================================================= #
     def _print_PyccelArraySize(self, expr):
-        return expr.fprint(self._print)
+        init_value = self._print(expr.arg)
```

Comment:
Did you delete the old fprint method?

Reply:
yes, you can check the new one in fcode.

Reply:
Yes I can see the new method. That is where I left the comment. But the old method (`PyccelArraySize.fprint`) is no longer needed. It is defined in ast/core.py but according to github that file hasn't been modified. Are you sure you pushed that change?

Reply:
oh, you're right i forgot a fprint in ast.core, it should be removed now.